### PR TITLE
Link Login Modal to follow a post

### LINKED ISSUE #4152: Link Login Modal to follow a post
We are creating a login/signup modal, next step is to link to various pages. In this issue we will be linking it to follow a post
link - https://publiclab.org/notes/joyofsoy/11-09-2018/fighting-the-flood-community-activism-and-education-in-pensacola
        `app/views/users/_follow.html.erb`

### LINKED ISSUE #4335: Link Login modal to blog tag page
We have a button to follow and unfollow on the tag/blog. We need to link it to the login modal if the person is not logged in.

### PR DESCRIPTION
**Pull request by dinaelhanan**
Fixes #4335 (<=== Add issue number here)

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below
* [ ]              ***ADD GIF**

### CODE REVIEW

**Comment by SidharthBansal** [comment]
on this issue I guess @okonek has supposed to work.
It is a quite lengthy issue and @okonek has proposed the solution for this. You may close this pr.

**Comment by SidharthBansal** [comment]
Sorry, for misunderstanding

**Comment by SidharthBansal** [comment]
Dina you need to do more in this issue.
Basic flows which you need to implement are
Not logged in->user logs in upon clicking button to follow->followed successfully and notification comes

Alternative flows
User has already followed the post. When user logs in they unfollow the post with flash message

File: app/views/tag/blog.html.erb [hunk 1]
```diff
@@ -25,7 +25,7 @@
         <% end %>
       <% else %>
         <div class="alert alert-info" role="alert">
-        <a href="/login?return_to=/blog" class="alert-link">Log in</a> to get updates for new posts!
+        <a data-toggle="modal" data-target="#loginModal?return_to=/blog">Log in</a> to get updates for new posts!
         </div>
```
Comment (dinaelhanan) [inline app/views/tag/blog.html.erb]: so would I add this:
@SidharthBansal 
```
 <% if current_user.following(params[:id]) %>
         <a rel="tooltip" title="<%= t('tag.blog.unfollow') %>" class="btn btn-default btn-sm active" href="/unsubscribe/tag/<%= params[:id] %>"><i class="fa fa-eye"></i> <%= t('tag.blog.following') %> <b><%= params[:id] %></b></a>
```
here? How would I make it so they unfollow when done login in? I know that the above handles  the unfollow, I just don't know how I would link it to here

**Comment by dinaelhanan** [comment]
@JonathanXu1 maybe you can also help

**Comment by JonathanXu1** [comment]
href="/unsubscribe/tag/<%= params[:id] %" 
^^ that's the link that would make someone unsubscribe

**Comment by dinaelhanan** [comment]
@JonathanXu1 ya, but I don't know how to make it so that when they finish login in, they are sent back to unfollow

**Comment by ValentinaTironi** [comment]
I think @kevinzluo did something like that with Javascript, maybe he can help here :)

**Comment by ValentinaTironi** [comment]
https://github.com/publiclab/plots2/pull/4228 check out this

**Comment by dinaelhanan** [comment]
@ValentinaTironi thanks for that! I'll try to adjust it to fit for this situation. Thanks!

**Comment by kevinzluo** [comment]
Hi @dinaelhanan 
What my code does is send a getjson request to the url so the request gets changed. The login modal always redirects to the latest request so that way they will go to the roght link :). Please just ping me if you need any help and I'll try to respond as soon as possible :+1:

**Comment by ValentinaTironi** [comment]
Dina, I think you are having this error in test because you are using `current_user` variable and in the test, you don't specific no one session
screen shot 2018-12-10 at 2 21 51 pm
Maybe I'm wrong, but try it you don't lose nothing

**Comment by dinaelhanan** [comment]
@ValentinaTironi thanks for that! I'll try to adjust it to fit for this situation. Thanks!

> Dina, I think you are having this error in test because you are using `current_user` variable and in the test, you don't specific no one session
>
I see that I am using the current_user variable, but I don't get what you mean by not specifying the no one session, could you explain more?

**Comment by ValentinaTironi** [comment]
Yes, of course. Check out this
https://github.com/publiclab/plots2/blob/d55049cf4953998b85a9a78722379f07fd42bef8/test/functional/admin_controller_test.rb#L38
Here, for example, you create a UserSession of Jeff user. So, when you put `current_user`, you are referring to Jeff. Do you understand?

**Comment by dinaelhanan** [comment]
so would I specify the current_user variable? I don't get how to fix it

**Comment by dinaelhanan** [comment]
so would I specify the current_user variable? Because if current_user is like Jeff, then the test should pass already.. right? @ValentinaTironi

**Comment by ValentinaTironi** [comment]
You have to create a user session, like here https://github.com/publiclab/plots2/pull/4225#issuecomment-445910842

**Comment by dinaelhanan** [comment]
@ValentinaTironi do I add that to the test file or here? (I'm guessing it's the test file, just want to make sure!)

**Comment by ValentinaTironi** [comment]
sure, test file :)

**Comment by dinaelhanan** [comment]
@ValentinaTironi wait like this?
```
test 'should make current_user the current user' do
    UserSession.create(users(:jeff))
    assert current_user = users(:jeff)
```

**Comment by dinaelhanan** [comment]
@ValentinaTironi reading over the fail error, it says that there's an undefined method 'following' I think _maybe_ I have to fix something there?

**Comment by dinaelhanan** [comment]
@ValentinaTironi so no assert in the test?

**Comment by ValentinaTironi** [comment]
nop

**Comment by dinaelhanan** [comment]
@ValentinaTironi ok I tried, let's see if it works. If not, I will add the assert then try again. and if that doesn't work, I'll call it a day and come back some other time >o<

**Comment by ValentinaTironi** [comment]
Oh, I'm gonna wait for travis checking and see what is happening :)

**Comment by dinaelhanan** [comment]
@ValentinaTironi got any ideas?

**Comment by SidharthBansal** [comment]
@ValentinaTironi @dinaelhanan please wait for some time. Someone is making a pr so that your work will be reduced.
Thanks everyone

**Comment by dinaelhanan** [comment]
@SidharthBansal I tried to look around but didn't find anything. Has anyone done some work relating to this?

**Comment by SidharthBansal** [comment]
Yeah we have Jonathan's PR. Waiting for it to be merged.

**Comment by SidharthBansal** [comment]
You can now continue to work on this issue.
#4902 is merged now. Please merge master into this branch before proposing your changes.

**Comment by dinaelhanan** [comment]
@SidharthBansal Ok I've merged master to the branch, but I'm having issues figuring this out, I'm getting errors in travis, and I don't know if I did it correctly. Thanks for all the help

**Comment by SidharthBansal** [comment]
What kind of errors?

File: test/functional/admin_controller_test.rb [hunk 2]
```diff
@@ -34,6 +34,10 @@ def teardown
     assert_redirected_to '/profile/' + user.username + '?_=' + Time.now.to_i.to_s
   end
 
+  test 'should make current_user the current user' do
```
[the remark this answers is not in the document]
Comment (SidharthBansal) [reply test/functional/admin_controller_test.rb]: Please remove these tests from here

**Comment by dinaelhanan** [comment]
@SidharthBansal these:
ERROR["test_browse_/blog", #<Minitest::Reporters::Suite:0x000055fe0655f190 @name="PublicPagesTest">, 14.572202021999942]
 test_browse_/blog#PublicPagesTest (14.57s)
ActionView::Template::Error:         ActionView::Template::Error: undefined method `following' for nil:NilClass
            app/views/tag/blog.html.erb:30:in `_app_views_tag_blog_html_erb___4233616707006234525_47274757633880'
            test/integration/public_pages_test.rb:47:in `block in <class:PublicPagesTest>'
ERROR["test_should_choose_i18n_for_tag/blog", #<Minitest::Reporters::Suite:0x000055fe03fb3c98 @name="I18nTest">, 63.004660034999915]
 test_should_choose_i18n_for_tag/blog#I18nTest (63.00s)
ActionView::Template::Error:         ActionView::Template::Error: undefined method `following' for nil:NilClass
            app/views/tag/blog.html.erb:30:in `_app_views_tag_blog_html_erb___4233616707006234525_47274757633880'
            test/integration/I18n_test.rb:428:in `block (2 levels) in <class:I18nTest>'
            test/integration/I18n_test.rb:424:in `each'
            test/integration/I18n_test.rb:424:in `block in <class:I18nTest>'

File: test/functional/admin_controller_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (SidharthBansal) [reply test/functional/admin_controller_test.rb]: Hi asked you to remove these changes.
Are you having any trouble removing these changes?

[the remark this answers is not in the document]
Comment (dinaelhanan) [reply test/functional/admin_controller_test.rb]: oops! I thought I made those changes.. I will do that now. Sorry!

**Comment by dinaelhanan** [comment]
@SidharthBansal OK I've removed them.  Sorry! I thought I made those changes earlier.

File: app/views/tag/blog.html.erb [hunk 3]
```diff
@@ -25,8 +25,13 @@
...
+	   <% if current_user.following(params[:id]) %>
```
[the remark this answers is not in the document]
Comment (dinaelhanan) [reply app/views/tag/blog.html.erb]: would I change it to `<% if current_user=nill %>` ?

[the remark this answers is not in the document]
Comment (SidharthBansal) [reply app/views/tag/blog.html.erb]: You should remove these changes.
You suppose to work for /tag/tagname. Here you are working for /tag/blog.
Hence, I have changed the issue number to a new issue created.

**Comment by dinaelhanan** [comment]
@SidharthBansal ok I've removed those changes. Thanks for the help!

**Comment by SidharthBansal** [comment]
>>I'd love to get started for #4335. 

You are solving #4335 instead of the issue which you have claimed. By mistake you jumped to the wrong file and found out an error which was similar & needs a fix.

**Comment by SidharthBansal** [comment]
Nice work @dinaelhanan tada! Thanks!
Merged :100:

**Comment by SidharthBansal** [comment]
You solved https://github.com/publiclab/plots2/issues/4335#issuecomment-448921085 login button instead of the issue on which you were supposed to work.
No problem. We found a new error and fixed it. Awesome!!! :tada: 
Thanks :balloon:  :balloon:

**Comment by dinaelhanan** [comment]
Working on fixing this on #4377

**Comment by SidharthBansal** [comment]
Moved to #4377

**Pull request merged**